web3_proxy/src/stats/stat_buffer.rs

```diff
@@ -127,22 +131,83 @@ impl StatBuffer {
                 stat = stat_receiver.recv() => {
                     // trace!("Received stat");
                     // save the stat to a buffer
+
+                    // TODO: tokio spawn this!
                     match stat {
-                        Some(AppStat::RpcQuery(stat)) => {
-                            if self.influxdb_client.is_some() {
-                                // TODO: round the timestamp at all?
+                        Some(AppStat::RpcQuery(request_metadata)) => {
+                            // we convert on this side of the channel so that we don't slow down the request
+                            let stat = RpcQueryStats::try_from_metadata(request_metadata)?;
+
+                            // update the latest balance
+                            // do this BEFORE emitting any stats
+                            let paid_credits_used: bool;
+                            let latest_balance: Balance;
+                            if let Some(db_conn) = self.db_conn.as_ref() {
+                                let user_id = stat.authorization.checks.user_id;
+
+                                // update the user's balance
+                                if user_id == 0 {
+                                    paid_credits_used = false;
+                                    latest_balance = Balance::default();
+                                } else {
+                                    // update the user's cached balance
+                                    let mut user_balance = stat.authorization.checks.latest_balance.write().await;
+
+                                    paid_credits_used = user_balance.active_premium();
+
+                                    user_balance.total_frontend_requests += 1;
```

Comment:
this seems extremely manual and bug-prone

Reply:
theres only one place that has a .write now. but we could move it into a function on Balance